codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -307,38 +338,87 @@ private Http2Stream requireStream(int streamId) {
         return stream;
     }
 
-    /**
-     * Wrap a DATA frame so it can be written subject to flow-control. Note that this implementation assumes it
-     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
-     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
-     * <p>
-     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
-     * the passed promise is not completed until last frame write.
-     * </p>
-     */
-    private final class FlowControlledData extends FlowControlledBase {
-        private ByteBuf data;
-        private int size;
-
-        private FlowControlledData(ChannelHandlerContext ctx, Http2Stream stream, ByteBuf buf, int padding,
-                                    boolean endOfStream, ChannelPromise promise) {
-            super(ctx, stream, padding, endOfStream, promise);
-            this.data = buf;
-            size = data.readableBytes() + padding;
+    private final class FlowControlledFileRegion extends FlowControlledData<FileRegion> {
+        
+        private FlowControlledFileRegion(ChannelHandlerContext ctx, Http2Stream stream,
+                FileRegion data, int padding, boolean endOfStream, ChannelPromise promise) {
+            super(ctx, stream, data, (int) data.count(), padding, endOfStream, promise);
```

Comment:
Yes, we need to deal with overflow. But I do not think we should throw exceptions. Transfer large files is an important usage of FileRegion I think. Maybe the Collection.size() is a good example-Return Integer.MAX_VALUE if actual size is overflow. Let me try.